modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java

```diff
@@ -1489,8 +1494,22 @@ private void execStoreOpcodes(StackFrame sf, int opcode, int[] operands) {
                     handleNullRefError();
                     break;
                 }
-
-                bMap.put(sf.stringRegs[j], sf.refRegs[k]);
+                BMapType mapType = (BMapType) bMap.getType();
+                if (sf.refRegs[k] == null) {
+                    if (!isValueType(mapType.getConstrainedType())) {
+                        // TODO improve logic for null case
```

Comment:
What are the cases we need to handle when the value is null?

Reply:
The problem here is for reference types we can't differentiate null cases as below.
Consider the below example.

map\<Person\> pMap = {};
updateMap(map m){
Employee e;
Person p;
//case 1
m["case-1"] = e;

//case 2
m["case-2"] = p;

//case 3
m["case-3"] = null;
}

All three cases ended up adding a null value with key.  Anyway we return null if key is not there for retrieval for ref types, so there should not be any issue. However key sets are different based on the decision whether to put null into ref map. That is why adding null is allowed for ref types. ( Value types there will be no issue because it will always be boxed when we put into  map  ) I put a todo because, if it is required to differentiate 2 cases as above, we may need to do it in here.